### PR TITLE
[FIXED-JENKINS-40826] Do not make network calls from the QueueListener thread

### PR DESCRIPTION
**Pull request by stephenc**
See [JENKINS-40826](https://issues.jenkins-ci.org/browse/JENKINS-40826)

@reviewbybees

### CODE REVIEW

File: src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubBuildStatusNotification.java [hunk 1]
```diff
@@ -176,18 +178,28 @@ private static GitHubSCMSource getSCMSource(final SCMSourceOwner scmSourceOwner)
      * With this listener one notifies to GitHub when a Job (associated to a PR) has been scheduled.
      * Sends: GHCommitState.PENDING
      */
-    @Extension public static class PRJobScheduledListener extends QueueListener {
+    @Extension
+    public static class PRJobScheduledListener extends QueueListener {
```
Comment (KostyaSha) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubBuildStatusNotification.java]: infinite fighting with jesse-styling :)

Comment (KostyaSha) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubBuildStatusNotification.java]: Btw, you don't need QueueListener at all.

Comment (stephenc) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubBuildStatusNotification.java]: People complain when I make big sweeping changes with 1000+ line diffs... they can see if they prefer death by 100 x 10 line diffs.

The point of the queue listener is to flag the status as early as possible. When we get the event from GH we do not know if any jobs will be triggered... we need to let the branch-api do its stuff and react to the event... that will queue jobs and then this listener flags the status before the job even starts a checkout

File: src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubBuildStatusNotification.java [hunk 2]
```diff
@@ -176,18 +178,28 @@ private static GitHubSCMSource getSCMSource(final SCMSourceOwner scmSourceOwner)
...
+            }
+            final long taskId = wi.getId();
+            final Job<?,?> job = (Job) wi.task;
+            // TODO would actually be better to use GitHubSCMSource.retrieve(SCMHead, TaskListener) (when properly implemented).
+            // That would allow us to find the current head commit even for non-PR jobs.
+            final SCMHead head = SCMHead.HeadByItem.findHead(job);
+            if (!(head instanceof PullRequestSCMHead)) {
+                return;
+            }
+            Computer.threadPoolForRemoting.submit(new Runnable() {
```
Comment (KostyaSha) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubBuildStatusNotification.java]: Why it called `Computer`? I thought it designed only for `remoting` communication for real `remoting` to slaves.

**Comment by rsandell** [comment]
:bee:

**Comment by rsandell** [comment]
:bee:

File: src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubBuildStatusNotification.java [hunk 3]
```diff
@@ -176,18 +178,29 @@ private static GitHubSCMSource getSCMSource(final SCMSourceOwner scmSourceOwner)
...
+            final long taskId = wi.getId();
+            final Job<?,?> job = (Job) wi.task;
+            // TODO would actually be better to use GitHubSCMSource.retrieve(SCMHead, TaskListener) (when properly implemented).
+            // That would allow us to find the current head commit even for non-PR jobs.
+            final SCMHead head = SCMHead.HeadByItem.findHead(job);
+            if (!(head instanceof PullRequestSCMHead)) {
+                return;
+            }
+            // TODO find a better general thread pool
+            Timer.get().submit(new Runnable() {
```
Comment (KostyaSha) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubBuildStatusNotification.java]: @stephenc it was only question. I'm searching the place where i can run own parallel (synchronised on some result things) things. Timer afair is limited to certain number of threads? And i remember you changed something in github-plugin.

Comment (KostyaSha) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubBuildStatusNotification.java]: SequentialExecutionQueue in github-plugin is on threadPoolForRemoting. So probably it right place?

**Pull request merged**